FeaturePoints/src/cz/fidentis/featurepoints/pdm/TrainingModel.java

```diff
@@ -0,0 +1,217 @@
+                valuesMatrix.set(j, 2, values.getFacialPoints().get(j).getPosition().z);
+            }
+
+            Matrix shapeSubtract = valuesMatrix.minus(meanShape);
+
+            covariance = covariance.plus(shapeSubtract.times(shapeSubtract.transpose()));
+        }
+
+        double[][] tmp = covariance.getArray();
+
+        if(trainingShapes.size() != 1){
+        
+        for (int j = 0; j < tmp.length; j++) {
+            for (int k = 0; k < tmp[j].length; k++) {
+                tmp[j][k] /= (trainingShapes.size() - 1);
+            }
+        }
+
+        }
+        return new Matrix(tmp);
+    }
+
+    /**
+     * return first i-th eigenVectors from covariance matrix based on eigenvalue, where value(i) >= value(i+1)
+     * @param i number of eigenVectors
+     * @param covariance
+     * @return first i-th matrix of eigenVectors
+     */
+    public Matrix getEigenVectors(int i, Matrix covariance) {
+        EigenvalueDecomposition eigen = covariance.eig();
+        double[][] vectors = eigen.getV().getArray();
+
+        return new Matrix(vectors);
+    }
+
+    /**
+     * return first i-th eigenValues from covariance matrix, where value(i) >= value(i+1)
+     * @param i number of eigen values
+     * @param covariance covariance matrix
+     * @return first i-th matrix of eigen values
+     */
+    public Matrix[] getEigenValues(int i, Matrix covariance) {
+        EigenvalueDecomposition eigen = covariance.eig();
+        Matrix[] matr = new Matrix[2];
+        double[] values = eigen.getRealEigenvalues();
+        double[][] values2 = eigen.getV().transpose().getArray();
+        sort(values, values2);
+
+        double[][] eigenValues = new double[1][i];
+        double[][] eigenVectors = new double[i][values2[0].length];
+        int tmp = 0;
+
+        for (int k = values.length - 1; k > values.length - 1 - i; k--) {
+
+            eigenValues[0][tmp] = values[k];
+            tmp++;
+        }
+        tmp = 0;
+        for (int k = values2.length - 1; k > values2.length - 1 - i; k--) {
+            for (int l = 0; l < values2[k].length; l++) {
+                eigenVectors[tmp][l] = values2[k][l];
+            }
+            tmp++;
+        }
+
+        matr[0] = new Matrix(eigenValues);
+        matr[1] = new Matrix(eigenVectors);
+
+        return matr;
+    }
+
+    /////// help sorting methods
+    
+    public double distancePoints(Vector3f point1, Vector3f point2) {
+        float x = point1.x - point2.x;
+        float y = point1.y - point2.y;
+        float z = point1.z - point2.z;
+
+        return Math.sqrt(x * x + y * y + z * z);
+    }
+
+    public void sort(double[] inputArr, double[][] inputArr2) {
+
+        if (inputArr == null || inputArr.length == 0) {
+            return;
+        }
+        array = inputArr;
+        array2 = inputArr2;
+        length = inputArr.length;
+        quickSort(0, length - 1);
+    }
+
+    private void quickSort(int lowerIndex, int higherIndex) {
```

Comment:
na sortovanie tiez existuje v module Utils trieda, prehod to prosimta tam, ak to pojde, urcite sa to zide aj inde